README.md
cmonk
=====

A simple ncurses-based frontend for Zabbix, Nagios, etc.


debian/changelog
cmonk (20190726-1) unstable; urgency=low

  * Minor UI fixes, added a config option to hide entries for some hosts/data 

 -- Rustam Tsurik <rustam@rustam-hsi>  Fri, 26 Jul 2019 12:07:10 +0300

cmonk (20190615-1) unstable; urgency=low

  * Updated zabbix plugin to make it compatible with Zabbix 4.x, code cleanup,
    etc.

 -- Rustam Tsurik <rustam.tsurik@gmail.com>  Sat, 15 Jun 2019 23:15:12 +0300

cmonk (20170525-1) unstable; urgency=low

  * Added on/off switch for modules in the yaml config

 -- Rustam Tsurik <rustam.tsurik@gmail.com>  Thu, 25 May 2017 13:05:41 +0300

cmonk (20170511-1) unstable; urgency=low

  * Refactoring, UI fixes

 -- Rustam Tsurik <rustam.tsurik@gmail.com>  Thu, 11 May 2017 13:36:12 +0300

cmonk (20161102-1) unstable; urgency=low

  * A quick and dirty fix, allow converting a non-reference into a json string.

 -- Rustam Tsurik <rustam.tsurik@gmail.com>  Wed, 02 Nov 2016 18:28:13 +0300

cmonk (20150928-1) unstable; urgency=low

  * A minor deb package dependencies fix.

 -- Rustam Tsurik <rustam.tsurik@gmail.com>  Mon, 28 Sep 2015 11:58:19 +0300

cmonk (20140127-1) unstable; urgency=low

  * Initial release 

 -- Rustam Tsurik <rustam.tsurik@gmail.com>  Mon, 27 Jan 2014 14:38:06 +0200


cMonkUI.pm
package cMonkUI;

use Curses;
use Switch;
use Time::Seconds;

=head1 NAME

cMonkUI - Curses UI module for cmonk

=head1 SYNOPSIS

    use cMonkUI;
    my $cui = cMonkUI->new();

    # Color scheme initialization. Draw the borders, etc.
    $cui->prepare(
        version => $VERSION,
        sort_order => $sort_order,
        display_columns => @display_columns,
    );

    # In a loop:
    # Clear the window, draw borders, etc.
    $cui->start_cycle(
        sort_order => $sort_order,
    );

    # For each monitoring plugin:
    $cui->print_header($running_threads{$thread_name}->{'name'});
    $cui->print_entry($host->{'hostname'}, $host->{'prio'}, $host->{'age'}, $host->{'data'});

    $cui->print_message($error_message);

    # Refresh screen
    $cui->finalize_cycle();

    # Detect screen resize, draw progress meter
    my $window_resized = $cui->pulse();

    # On exit
    $cui->cleanup();



=head1 DESCRIPTION

This module is supposed to be used as a part of cmonk software.

=head2 Methods

=cut

=item C<cMonkUI::prepare()>

Color scheme initialization. Draw the borders, etc.
This method is called only once during the program initialization.

=cut

sub prepare {
    my ($self, %args) = @_;

    initscr();
    curs_set(0);
    $win = Curses->new();
    $self->{'window'} = $win;

    $self->{'window'}->getmaxyx($ymax, $xmax);

    $self->{'x-max'} = $xmax;
    $self->{'y-max'} = $ymax;

    $self->{'pulse'} = 1;

    start_color();
    init_pair(1, COLOR_GREEN, COLOR_BLACK);
    init_pair(2, COLOR_RED, COLOR_BLACK);
    init_pair(3, COLOR_WHITE, COLOR_BLACK);
    init_pair(4, COLOR_YELLOW, COLOR_BLACK);
    init_pair(5, COLOR_MAGENTA, COLOR_BLACK);

    $win->attron(COLOR_PAIR(3));
    $win->box(0,0); $win->attroff(COLOR_PAIR(3));

    $win->attron(COLOR_PAIR(1)); 
    $win->addstr(1, 1, "Preparing data...");
    $win->attroff(COLOR_PAIR(1));


    $self->{'display_columns'} = $args{'display_columns'};

    $self->{'version'} = $args{'version'};
    $self->{'window'}->addstr(
        $self->{'y-max'} - 1, 1,
        "v" . $self->{'version'} . " | " . $args{'sort_order'} . " "
    );

    $win->refresh();
}

# -- cMonkUI::cleanup()
sub cleanup(){
    my ($self, %args) = @_;
    curs_set(1);
    endwin();
}

=item C<cMonkUI::start_cycle($self, %args)>

Prepare for screen redraw.

=cut

sub start_cycle {
    my ($self, %args) = @_;

    my $xmax; my $ymax;
    $self->{'window'}->getmaxyx($ymax, $xmax);

    $self->{'x-max'} = $xmax;
    $self->{'y-max'} = $ymax;
    $self->{'y-pos'} = 1; 

    $win->clear();

    $win->attron(COLOR_PAIR(3));
    $win->box(0,0);
    $win->attroff(COLOR_PAIR(3));

    $self->{'window'}->addstr(
        $self->{'y-max'} - 1, 1,
        "v" . $self->{'version'} . " | " . $args{'sort_order'} . " "
    );
}

=item C<cMonkUI::finalize_cycle()>

Redraw the screen.

=cut

sub finalize_cycle {
    my ($self, %args) = @_;

    $self->{'window'}->refresh();
}

=item C<cMonkUI::pulse()>

Detect screen resize / draw a progress meter.

=cut

sub pulse {
    my ($self, %args) = @_;
    my $resize = 0;

    my $xmax; my $ymax;
    $self->{'window'}->getmaxyx($ymax, $xmax);
    if (($self->{'x-max'} != $xmax) | ($self->{'y-max'} != $ymax)) {
        $self->{'x-max'} = $xmax;
        $self->{'y-max'} = $ymax;
        $resize = 1;
    }

    my $puchr = '/';
    if ($self->{'pulse'} == 2 ) { $puchr = '-'; }
    if ($self->{'pulse'} == 3 ) { $puchr = "\\"; $self->{'pulse'} = 0; }
    $self->{'pulse'} += 1;

    $self->{'window'}->addstr(1, $xmax - 3, $puchr);
    $self->{'window'}->refresh();

    return $resize;
}

=item C<color_ctl($self, $priority, $operation)>

Set text color accorging to the priority

=cut

sub color_ctl {
    my ($self, $priority, $operation) = @_;

    my %color_table = (
        "low"    => 3,  # white
        "high"   => 2,  # red
        "medium" => 4,  # yellow
    );

    my $color_pair = $color_table{$priority};

    switch ($operation) {
        case ('on') { $self->{'window'}->attron(COLOR_PAIR($color_pair)) }
        case ('off') { $self->{'window'}->attroff(COLOR_PAIR($color_pair)) }
    }
}

=item C<cMonkUI::print_header($self, $text)>

Output a header for a monitoring plugin.

=cut

sub print_header {
    my ($self, $text) = @_;

    # Add some space.
    $self->{'y-pos'} += 1;

    $win->attron(COLOR_PAIR(1));
    $self->{'window'}->addstr($self->{'y-pos'}, 1, substr($text, 0, $self->{'x-max'} - 2));
    $win->attroff(COLOR_PAIR(1));

    $self->{'y-pos'} += 1;
}

# -- cMonkUI::print_entry
# -- Output data for a host/trigger.
sub print_entry {
    my ($self, $host, $prio, $age, $data) = @_;

    if ($host ne 'na') {

        # convert age to pretty format
        my $age_ts=Time::Seconds->new($age);
        my $ap = $age_ts->pretty;
        $ap =~ tr/,//d;
        $ap =~ s/ seconds/sec/;
        $ap =~ s/ minutes/min/;
        $ap =~ s/ hours/hr/;
        $ap =~ s/ days/d/;

        my $text = "";
        
        my $col_counter = 1;
        my $col_max = scalar @{$self->{'display_columns'}};
        
        # output the data
        foreach $col (@{$self->{'display_columns'}}) {
            if ($col eq 'hostname') {
                #my $text = "$host - ($ap) - $data";
                $text .= $host;
            }
            if ($col eq 'age') {
                $text .= $ap;
            }
            if ($col eq 'data') {
                $text .= $data;
            }

            # delimiter        	
            if ($col_counter++ < $col_max){
                $text .= " | ";    
            }
            
        }
        
        $self->color_ctl($prio, "on");
        $self->{'window'}->addstr($self->{'y-pos'}, 1, substr($text, 0, $self->{'x-max'} - 2));
        $self->color_ctl($prio, "off");

        $self->{'y-pos'} += 1;
    }
}

=item C<cMonkUI::print_entry($self, $text)>

Output a text message.

=cut

sub print_message {
    my ($self, $text) = @_;

    $self->color_ctl("low", "on");
    $self->{'window'}->addstr($self->{'y-pos'}, 1, substr($text, 0, $self->{'x-max'} - 2));
    $self->color_ctl("low", "off");

    $self->{'y-pos'} += 1;
}


=item C<cMonkUI::new()>

The class constructor.

=cut

sub new {
    my ($class, %args) = @_;
    bless \%args, $class;
    return $class;
}

1;

__END__

=head1 AUTHOR

Rustam Tsurik E<lt>rustam.tsurik@gmail.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2013-2019 by Rustam Tsurik

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.0 or,
at your option, any later version of Perl 5 you may have available.

=cut

modules/nagios.pm
# -- Nagios cMonk plugin
# -- Copyright (C) 2013 Rustam Tsurik

{
  package NagiosParse;
  use base "HTML::Parser";
  use URI::Escape;

  # -- Init vars.
  my $capture = 0;
  my $table_lock = 0;
  my $td_counter = 0;

  my $wrk_host = "na";
  my $wrk_serv = "na";
  my $wrk_age = 0;
  my $wrk_prio = "low";
  my $wrk_text = "No data";


  # -- NagiosParse::end()
  # -- This is not called directly, matches tags end
  sub end {
    my ($self, $tag, $origtext) = @_;

    if ($tag =~ 'table'){
      if ($table_lock > 0){ $table_lock-- }
    }
  }

  # -- NagiosParse::start()
  # -- This is not called directly, matches tags start
  sub start {
    my ($self, $tag, $attr, $attrseq, $origtext) = @_;

    if ($tag =~ 'table'){
        if (($origtext =~ "CLASS=\'status\'")|($origtext =~ "class=\'status\'")){
            $table_lock = 0;
        } else { $table_lock++ }
    }
    # ------ process TR tag
    if (($tag eq 'tr')&(!$table_lock)){
        $td_counter = 0;    # <td> counter = 0
    }
    if (($tag eq 'td')&(!$table_lock)){
        $td_counter++;
        $capture = 1;
    }
    if ($tag eq 'a'){
        if ($origtext =~ "extinfo\.cgi(.*)&host=(.*)&service=(.*)\'>"){
            $wrk_host = $2;
            $wrk_serv = uri_unescape($3);
            $wrk_serv =~ tr{\+}{\ }d;
            $wrk_prio = 'low';
        }
    }
  }

  # -- NagiosParse::text()
  # -- This is not called directly, matches tags contents
  sub text {
    my ($self, $text) = @_;

    if (($capture == 1)&($td_counter>0)&($td_counter<8)){
        $text =~ tr{\n\r}{}d;

        my %table_columns = (
        1 => sub {},
        2 => sub {},
        3 => sub { 
               if ($text =~ m/WARN/) { $wrk_prio = 'medium'};
               if ($text =~ m/CRITICAL/) { $wrk_prio = 'high'};
             },
        4 => sub {}, # here $text == last check, don't need this.
        5 => sub { 
               $text=~ s/^\s//; # <-- age, convert to seconds + adjust according to current time.
               $text =~ s/[\sdhms]+/ /g;
               my @duration_parts = split / /, $text;
               $wrk_age =  $duration_parts[0]*86400 + $duration_parts[1]*3600 + $duration_parts[2]*60 + $duration_parts[3];
             },
        6 => sub {}, # here $text == attempt, don't need this.
        7 => sub { 
               $text =~ s/\&.([0-9]+)\;/pack('C',$1)/ge; 
               $text =~ s/&nbsp;/ /g;

               # push host data into array here
               # ...
               if (($wrk_prio eq 'high') | ($wrk_prio eq 'medium')) {
                 my $c = $self->{cnt};
                 push (
                   @{$self->{data}}, 
                   { 
                     "hostname" => $wrk_host, 
                     "data" => $wrk_serv . ' - ' . $text,
                     "priority" => $wrk_prio,
                     "age" => $wrk_age,
                   }
                 );
                 $self->{cnt} = ($c + 1);
               }
             }
        );

        $table_columns{$td_counter}->();
        $capture = 0;
    }
  }

  # -- NagiosParse::reset_counter()
  # -- Init the class variables before use
  sub reset_counter {
    my ($self, $value) = @_;

    $self->{cnt} = 0;
    $self->{data} = ();
  }

  # -- NagiosParse::get_hosts()
  # -- Returns the collected data
  sub get_hosts {
    my ($self, %value) = @_;

    # -- return a reference to the data array
    my @ret_hosts = $self->{data};
    return \@ret_hosts; 
  } 
} 
# -- package NagiosParse ends here

{
  package nagios;
  use LWP::UserAgent;
  use JSON;
  use Encode;
  use Data::Dumper;

  # -- nagios::get_data()
  # -- Retrieve the web page, parse it and return json code. 
  sub get_data {
    my ($op, $conf, $session) = @_;
  
    if ($op eq 'logout') {
      return 1; 
    }
    
    if ($op eq 'login') {
      return '';
    }


    $nag_usr = $conf->{'api_user'};
    $nag_pwd = $conf->{'api_pass'};

    my $myuseragent = new LWP::UserAgent;
    $myuseragent->timeout(120);

    # -- Disable SSL verification.

    my $sslver = "on";
    if (defined $conf->{"verify_ssl"}) {
      $sslver = $conf->{"verify_ssl"};
    }

    if ($sslver eq "off") {
      $myuseragent->ssl_opts(
        verify_hostname => 0,
        SSL_verify_mode => 0x00
      );
    }


    my($uri_scheme, $hostname, $uri_path) =
    $conf->{'api_uri'} =~ m|(https?://)([^/]*)(.*)|;

    #my $url = $conf->{'api_uri'} . "/cgi-bin/status.cgi?host=all&limit=0";

    # retrieve issues only
    my $url = $conf->{'api_uri'} . '/cgi-bin/status.cgi?host=all&limit=0&servicestatustypes=28&hoststatustypes=15';

    my $auth_realm = 'Nagios Access';
    if (defined $conf->{'api_realm'}){
      $auth_realm = $conf->{'api_realm'};
    }

    $myuseragent->credentials($hostname . ":80", 
      $auth_realm , $nag_usr => $nag_pwd);

    $myuseragent->credentials($hostname . ":443",
      $auth_realm , $nag_usr => $nag_pwd);


    my $request = new HTTP::Request('GET', $url);
    my $response = $myuseragent->request($request);
    
    my $json = new JSON; $json->allow_nonref(1);
    my $json_text; 

    if (!$response->is_success) {
      # -- Couldn't retrieve the web page.
      $json_text = $json->encode({ "status" => "error", "errormsg" => "Couldn't retrieve data." });

    } else {
      # -- Web page retrieved, parse it.
      my $content = decode_utf8 $response->content();

      my $dash_parser = NagiosParse->new();
      $dash_parser->reset_counter();
      $dash_parser->parse($content);
      my @ext_parsed = @{$dash_parser->get_hosts()};

      $json_text = $json->encode({ "status" => "ok", "session" => "", "data" => @ext_parsed });

      # -- clean up
      $dash_parser->reset_counter();
      @ext_parsed = ();
    }

    # -- return JSON code
    return $json_text;
  }

  # -- Register the plugin.
  # -- This is executed when module loads.
  $main::modules_callbacks{'nagios'} = \&get_data;
}

1;


modules/zabbix.pm
package zabbix;

use strict;

use JSON;
use Switch;

use JSON::RPC::Legacy::Client;
use Time::Seconds;

my $_module_name = 'zabbix';
my $_module_version = '1.0.0';

=head1 NAME

zabbix - Zabbix plugin for cmonk

=head1 SYNOPSIS

    our %modules_callbacks = ();
    require modules/zabbix.pm;
    my $plugin_json_output = $modules_callbacks{'zabbix'}->('get', $module_config, $session);

=head1 DESCRIPTION

This module is supposed to be used as a part of cmonk software.

=head2 Methods

=cut

# do_rpc_call() puts error messages into these vars,
# this needs to be rewritten.
my $rpc_error = 0;
my $rpc_error_msg = "";

=item C<do_rpc_call($uri, $request, ssl_verification)>

This is an internal method that performs RPC calls 
and is to be used by other module methods. 

=cut 

sub do_rpc_call {
    my (
        $uri, $req, 
        $ssl_verification
    ) = @_;

    my $client = new JSON::RPC::Legacy::Client;

    # Disable SSL verification if requested.
    if ($ssl_verification eq "off") {
        $client->ua->ssl_opts(
            verify_hostname => 0,
            SSL_verify_mode => 0x00
        );
    }

    # Run the API call.
    my $res = $client->call($uri, $req);

    # Check if we got an error and put it in the global vars.
    if ($res) { 
        if ($res->is_error) { 
            $rpc_error = 1; 
            $rpc_error_msg = $res->content->{'error'}->{'data'};
        }
    } else { 
        # We didn't even get an error code.
        $rpc_error = 1; $rpc_error_msg = "Unknown error";
    }

    # If we have any content returned by API call, return it.
    if (defined $res->content) {
        return $res->content->{'result'};
    } else { return () }
}

=item C<zabbix_user_login($module_config, $session_token)>

This method authenticates to the Zabbix server and returns a session token.
It is to be used by other module methods, e.g. zabbix_get_data().

The login info is passed in the module config object.

    $module_config->{'api_user'} = 'api';
    $module_config->{'api_pass'} = 'myPassword';

=cut

sub zabbix_user_login {
    my ($module_config, $session_token) = @_;

    my $req = {
        method  => 'user.login',
        id => 0,
        jsonrpc => "2.0",
        params  => {
            user => $module_config->{'api_user'},
            password => $module_config->{'api_pass'}
        },
    };

    my $new_session_token = do_rpc_call(
        $module_config->{'api_uri'}, $req, $module_config->{'verify_ssl'}
    );
    return $new_session_token;
}

=item C<zabbix_user_logout($module_config, $session_token)>

This method logs out from an existing session on the Zabbix server.
Not to be used directly.

=cut

sub zabbix_user_logout {
    my ($module_config, $session_token) = @_;

    # We don't have a session token, exit  
    if ($session_token eq "") {
        return;
    }
 
    # https://www.zabbix.com/documentation/4.2/manual/api/reference/user/logout
    my $req = {
        jsonrpc => "2.0",
        method => 'user.logout',
        auth => $session_token,
        params => [],
        id => 8,
    };

    my $response = do_rpc_call(
        $module_config->{"api_uri"}, $req, $module_config->{'verify_ssl'}
    );

    if ($response == JSON::true ) {
        return 1;
    }
}


=item C<zabbix_get_data($module_config, $session_token)>

This method retrieves all current problems(active triggers).

=cut

sub zabbix_get_data {
    my ($module_config, $session_token) = @_;

    my $new_session_token;    

    # Check if we already have a saved auth key.

    if ($session_token ne "") {
        $new_session_token = $session_token; 
    } else {
        $rpc_error = 0;

        $new_session_token = zabbix_user_login( 
            $module_config, $session_token
        ); 

        if ($rpc_error == 1) {
            # Authentication error
            my $json_text = JSON->new()->allow_nonref(1)->encode(
                { 
                    "status" => "error",
                    "errormsg" => "AUTH ERROR: $rpc_error_msg",
                    "session" => "",
                    "data" => ""
                }
            );
            return $json_text;
        }
    }

    # Request to get triggers.
    my $req = {
        'method'  => 'trigger.get',
        'auth' => $new_session_token,
        'id' => '3',
        'jsonrpc' => '2.0',
        'params'  => {
            'output' => 'extend',
            'select_hosts' => 'extend',
            'sortfield' => 'lastchange',
            'sortorder' => 'ASC',
            'expandExpression' => '0',
            'expandDescription' => '1',
            'selectHosts' => ['hostid', 'name'], 
            'active' => '1',
            'filter' => {
              'value' => '1',
            }
        }
    };

    # Attempt to retrieve triggers.
    $rpc_error = 0;
    my $triggers_data = do_rpc_call(
        $module_config->{'api_uri'}, $req, 
        $module_config->{'verify_ssl'}
    );

    if ( ($rpc_error == 1) && ($rpc_error_msg eq 'Not authorized') ) {
        # Not authorized, try to re-auth.
        $new_session_token = zabbix_user_login( $module_config, $session_token); 
        $req->{'auth'} = $new_session_token;

        # And then run the trigger.get call once again
        $rpc_error = 0;
        $triggers_data = do_rpc_call( 
            $module_config->{'api_uri'}, $req, 
            $module_config->{'verify_ssl'} 
        );
    }

    # Re-auth didn't help, or initially got some other error.
    if ($rpc_error == 1) {
        return JSON->new()->encode(
            { 
                'status' => 'error', 
                'errormsg' => "ERROR: $rpc_error_msg", 
                'session' => '', 
                'data' => '' 
            }
        );
    }

    my @data = ();

      foreach my $trigger (@{$triggers_data}) {
        # This call may select multiple hosts, 
        # use only the first one.
        my $hostname = $trigger->{'hosts'}->[0]->{'name'};

        my $description = $trigger->{'description'};
        my $age_in_seconds = (time - $trigger->{'lastchange'});

        # Just in case we got a negative age, but this shouldn't happen
        if ($age_in_seconds < 0) { $age_in_seconds = 0 };

        # Zabbix has five levels of priority, we convert them down to three;
        # 5 - disaster; 4 - high; 3 - average; 2 - warning; 1 - info
        my $priority = 'low';

        if ( $trigger->{'priority'} < 3 ) { 
            $priority = 'low' 
        } elsif  ( $trigger->{'priority'} == 3) { 
            $priority = 'medium' 
        } elsif ( $trigger->{'priority'} > 3) {
            $priority = 'high'
        }

        # Push the parsed data into the array
        push ( @data, 
            { 
              "hostname" => $hostname, 
              "data" => $description,
              "priority" => $priority,
              "age" => $age_in_seconds
            }
        );
    }

    return JSON->new()->allow_nonref(1)->encode(
        { 
            "status" => "ok",
            "session" => $new_session_token,
            "data" => \@data 
        }
    );
}

=item C<_module_callback($call_name, $module_config, $session_token)>

This method will be registered in the main cmonk program 
as a callback for the zabbix module.

The call name can be "get", "login" or "logout". 

=cut

sub _module_callback {
    my ( $call_name, $module_config, $session_token ) = @_;

    switch ($call_name) {
        case 'login'  { 
            return zabbix_user_login( $module_config, $session_token );
        }
        case 'logout' {
            return zabbix_user_logout( $module_config, $session_token ); 
        }
        case 'get'    { 
            return zabbix_get_data( $module_config, $session_token );
        }
    }
}

# Insert this plugin into the callbacks array in the main module
$main::modules_callbacks{ $_module_name } = \&_module_callback;

1;

__END__

=head1 AUTHOR

Rustam Tsurik E<lt>rustam.tsurik@gmail.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2013-2019 by Rustam Tsurik

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.0 or,
at your option, any later version of Perl 5 you may have available.

=cut


cmonk-cui.pl
#!/usr/bin/perl

#
# cmonk - console monitoring kit 
# A simple ncurses-based frontend for Zabbix, Nagios, etc.
# Version 2.0.2, 2019-06-15
# Copyright (C) 2013-2019 Rustam Tsurik
# 

use strict;
use warnings;

use Term::ANSIColor qw(:constants);  # Do we use this module?
local $Term::ANSIColor::AUTORESET = 1;

use threads;
use threads::shared;
use Switch;
use JSON;
use Term::ReadKey;
use YAML::XS;
use Env qw(HOME);
use Time::HiRes qw(usleep nanosleep);
use File::Basename;

my $VERSION = '2.0.2';

my $TICK_PERIOD = 100000;        # 100'000us = 10 ticks per second
my $TICK_MAX = 10;                # Reset UI every 10 ticks

my $user_config;                # Loaded user config(.cmonk.yaml) as a Perl object

our %modules_callbacks = ();    # Shared with modules, contains the list of callbacks subs 
my %running_threads = ();        # The list of running threads, includes names and handles

my $do_ui_redraw :shared;        # Used to send the redraw signal from threads to the main process
my $do_shutdown :shared;        # Used to tell the threads to exit the loop
my %last_data :shared;            # Data received from the modules

my $sort_order = 'age';            # Issues sort order, e.g. age or hostname

## start_thread() - The thread subroutine
## arguments: TODO

sub start_thread {

    my ($thread_name, $module_config) = @_;

    # Setup a handler for shutting down this thread
    local $SIG{'KILL'} = sub {
        threads->exit(); 
    };
    
    # The default loop timeout is 60 seconds,
    # but if we have a user-defined one, use it instead
    my $loop_timeout = 60;
    if (defined $module_config->{'refresh'}) {
        $loop_timeout = $module_config->{'refresh'};
    }
    
    # The module type, e.g. zabbix, nagios
    my $module_type = $module_config->{'type'};

    # Enter the loop
    while ($do_shutdown == 0) {

        # Retrieve session data from the last data
        my $json = JSON->new(); $json->allow_nonref(1);
        my $thread_data = $json->decode($last_data{$thread_name});
        my $sess = $thread_data->{'session'};

        # Exec the callback function
        # Send the complete module config and session data
        my $plugin_json_output = $modules_callbacks{$module_type}->('get', $module_config, $sess);

        # Push new data for this thread into the shared variable.
        {
            lock(%last_data);
            $last_data{$thread_name} = $plugin_json_output;
        };

        # Got new data, force UI redraw
        {
            lock($do_ui_redraw);
            $do_ui_redraw = 1;
        };

        sleep($loop_timeout);
    }

}

## load_configs() - Loop through the list of dirs(.:$HOME:/etc), find and load the config
## arguments: NONE

sub load_configs {

    my @CONFIG_DIRS = ('.', $HOME, '/etc');
    my $user_config_path; 

    foreach my $dir (@CONFIG_DIRS) {
        $user_config_path = "$dir/.cmonk";
        (-e $user_config_path) && last;
    }

    die "Couldn't find the config file!" unless (-e $user_config_path);
    $user_config = YAML::XS::LoadFile( $user_config_path );

    (defined $user_config->{'modules'}) or die("Config file parse error: the modules section not found");

    # Default sort order is age, see at the top.
    if (defined $user_config->{'gui'}->{'sort_order'}){
        $sort_order = $user_config->{'gui'}->{'sort_order'};
    }    

}

## load_modules() - TODO
## arguments: NONE

sub load_modules {

    my $modules = $user_config->{'modules'};
    my %loaded_modules = ();

    my $thread_id = 0;

    foreach my $mod (@$modules) {
        # Is it already loaded?
        if (! defined $loaded_modules{$mod->{'type'}} ) {

            # Nope, build the path to this module and attempt to load. 
            my $new_module = "modules/" . $mod->{'type'} . ".pm";
            eval { 
                require $new_module; 
            };
        
            # Check if it was actually loaded and register as such.
            (defined $INC{$new_module}) or die("Unable to load the plugin for " . $mod->{'type'} ."\n");
            $loaded_modules{$mod->{'type'}} = 1;
        }

        # Check whether this module is disabled in the config
        my $mod_enabled = 1;
        if (defined $mod->{'enabled'}){
            if ($mod->{'enabled'} eq 'off') { $mod_enabled = 0; }
        }

        if ($mod_enabled){
            # Now spawn a thread for this module
            # Build a name for this thread, e.g. 10-zabbix-2        
            my $thread_name = $mod->{'position'} . '-' . $mod->{'type'} . "-". $thread_id;

            $running_threads{$thread_name} = {
                "name" => $mod->{'name'},
                "handle" => threads->create('start_thread', $thread_name, $mod),
                "mod" =>  $mod,
            };

            # Init JSON data for this thread
            #
            # data => monitoring data
            # session => auth/session data for this module
            # status => ok/error
            # errormsg => the message
            $last_data{$thread_name} = 
            '{"data":null,"session":"","status":"error","errormsg":"Waiting for data..."}'; 
        }
    } 

}

sub cleanup_and_exit {
    my ($cui) = @_;

    $cui->cleanup();        # clean up UI

    ReadMode('normal');

    # It's likely that this var is already set to 1, but anyway.
    $do_shutdown = 1;

    print "Logging out of all instances, please wait.\n";

    # Now, shut down all threads by sending the KILL signal
    foreach my $thread_name (keys %running_threads) {
        print "- " . $running_threads{$thread_name}->{'name'};

        $running_threads{$thread_name}->{'handle'}->kill('KILL')->detach();

        # The module type, e.g. zabbix or nagios, and session token
        my $module_type = $running_threads{$thread_name}->{'mod'}->{'type'};
        my $sess = JSON->new->allow_nonref->decode( $last_data{$thread_name} )->{'session'};

        # Exec the callback function
        # Send the complete module config and session data
        my $plugin_json_output = $modules_callbacks{$module_type}->(
            'logout', $running_threads{$thread_name}->{'mod'}, $sess
        );

        print " [done]\n";
    }

    exit(0);
}

sub setup_ui {

    my ($cui) = @_;

    my @display_columns = ['hostname', 'age', 'data'];
    if (defined $user_config->{'gui'}->{'display_columns'}){
        @display_columns = $user_config->{'gui'}->{'display_columns'};
    }

    ReadMode('cbreak');
    $cui->prepare(
        version => $VERSION,
        sort_order => $sort_order,
        display_columns => @display_columns,
    );
}

sub check_keypress {

    my $read_key = ReadKey(-1); # -1 == non-blocking read
    if (defined $read_key) {

        switch ($read_key) {
            case ["q", "Q"]    {
                # Shutdown the app
                $do_shutdown = 1;
            }
            case ["s", "S"]    {
                # Change the sort order
                if ($sort_order eq 'age'){
                    $sort_order = 'hostname';
                } else {
                    $sort_order = 'age';
                }

                # and force redraw
                {
                    lock $do_ui_redraw;
                    $do_ui_redraw = 1;
                }
            }
        }

    }
}

sub redraw_ui {
    
    my ($cui) = @_;

    # Clear the window, draw borders, etc
    $cui->start_cycle(
        sort_order => $sort_order,
        );

    foreach my $thread_name (sort keys %last_data) {
        # Print the thread name.
        $cui->print_header($running_threads{$thread_name}->{'name'});

        my $json = new JSON; $json->allow_nonref(1);
        my $thread_data = $json->decode($last_data{$thread_name});
        
        if ($thread_data->{'status'} eq 'ok'){
            my $count = 0;

            # Reorder data as per settings
            my @sorted_thread_data = ();

            # won't sort if empty
            if (scalar $thread_data->{'data'} > 0) {
                if ($sort_order eq 'age'){
                    @sorted_thread_data = sort {$b->{'age'} <=> $a->{'age'}} @{$thread_data->{'data'}}
                } else {
                    @sorted_thread_data = sort {$a->{'hostname'} cmp $b->{'hostname'}} @{$thread_data->{'data'}}
                }

                foreach my $entry ( @sorted_thread_data ) {
                    # Print the entry info
                    my $skip = 0;

                    if (defined $user_config->{'gui'}->{'hide_data'}) {
                        if ( $entry->{'data'} =~ $user_config->{'gui'}->{'hide_data'} ) {
                            $skip = 1;
                        }
                    }

                    if (defined $user_config->{'gui'}->{'hide_hostname'}) {
                        if ( $entry->{'hostname'} =~ $user_config->{'gui'}->{'hide_hostname'} ) {
                            $skip = 1;
                        }
                    }

                    if ($skip == 0){
                        $cui->print_entry(
                            $entry->{'hostname'}, 
                            $entry->{'priority'},
                            $entry->{'age'}, 
                            $entry->{'data'}
                            );
                        $count++;
                    }
                }
            }
            if ($count == 0) {
                $cui->print_message("No issues detected.");
            }
        } else {
            my $error_message = $thread_data->{'errormsg'};
            $cui->print_message($error_message);
        }

        # Done with GUI, check threads health and restart if needed.
        if (! $running_threads{$thread_name}->{'handle'}->is_running()) {
            $cui->print_message("Thread is NOT running, restart attempted.");

            my $mod = $running_threads{$thread_name}->{'mod'};
            $running_threads{$thread_name}->{'handle'} = threads->create('start_thread', $thread_name, $mod);
        }
    }

    # Refresh the window
    $cui->finalize_cycle();

    # Done with redrawing
    { 
        lock($do_ui_redraw); 
        $do_ui_redraw = 0; 
    }
}

## 
## Let's go! Start the main thread.
##

# Init the variables
$do_shutdown = 0;    # Global, shared var
$do_ui_redraw = 0;    # Global, shared var

# Load the configs and load the modules
load_configs();
load_modules();

# Create a new UI object and call the setup_ui function
my $cui = cMonkUI->new();
setup_ui($cui);

# Set up a handler for the INT signal/Ctrl-C
$SIG{'INT'} = sub { 
    cleanup_and_exit($cui);
};

# Main loop tick counter
my $tick = 1;

# Enter the main loop
while ($do_shutdown == 0) {

    # Check if any keys have been pressed
    check_keypress();

    # Redraw UI if we got new data or something
    if ($do_ui_redraw == 1) {
        redraw_ui($cui);
    }

    usleep($TICK_PERIOD);
    if ($tick == $TICK_MAX) {
        $tick = 1;

        # cMonkUI->pulse() updates the progress bar and detects window resize
        my $window_resized = $cui->pulse();
        if ( $window_resized == 1 ) {
            lock($do_ui_redraw); 
            $do_ui_redraw = 1; 
        }
    }
    $tick++;
}

# We are done, clean up and exit
cleanup_and_exit($cui);

# Load the UI module
BEGIN {
    my $dir = dirname(__FILE__);
    push @INC, $dir; 
    require cMonkUI;
}
